### PR TITLE
chore: group & release events for local eventbus

### PR DESCRIPTION
**Pull request by riqwan**
what:

- adds ability to group, release and clear events on local eventbus

RESOLVES CORE-2262

### CODE REVIEW

File: packages/core/utils/src/event-bus/index.ts [hunk 1]
```diff
@@ -24,6 +24,15 @@ export abstract class AbstractEventBusModuleService
   abstract emit<T>(data: EventBusTypes.EmitData<T>[]): Promise<void>
   abstract emit<T>(data: EventBusTypes.Message<T>[]): Promise<void>
 
+  abstract stageEvent<TData = unknown>(
```
Comment (adrien2p) [inline packages/core/utils/src/event-bus/index.ts]: suggestion: should probably support the same API as emit no?

File: packages/modules/event-bus-local/src/services/__tests__/event-bus-local.js [hunk 2]
```diff
@@ -65,6 +65,124 @@ describe("LocalEventBusService", () => {
...
+
+        eventBus.emit("test-event", { test: "1234", eventGroupId: "test" })
+
+        expect(eventBus.eventEmitter_.emit).not.toHaveBeenCalled()
+        expect(stageEventFn).toHaveBeenCalledTimes(1)
+        expect(stageEventFn).toHaveBeenCalledWith("test", "test-event", {
+          test: "1234",
+        })
+
+        eventBus.emit("test-event", { test: "1234", eventGroupId: "test" })
```
Comment (adrien2p) [inline packages/modules/event-bus-local/src/services/__tests__/event-bus-local.js]: suggestion: can we also emit without group id to ensure the behaviour is fully tested ?

**Review by adrien2p — APPROVED** [review]
LGTM, few suggestions still

File: packages/modules/event-bus-local/src/services/event-bus-local.ts [hunk 3]
```diff
@@ -70,8 +74,56 @@ export default class LocalEventBusService extends AbstractEventBusModuleService
...
+      await this.stageOrEmitEvent(event.eventName, data)
+    }
+  }
+
+  // If the data of the event consists of a eventGroupId, we don't emit the event, instead
+  // we add them to a staging queue grouped by the eventGroupId and release them when
+  // explicitly requested.
+  // This is useful in the event of a distributed transaction where you'd want to emit
+  // events only once the transaction ends.
+  async stageOrEmitEvent(
```
Comment (carlos-r-l-rodrigues) [inline packages/modules/event-bus-local/src/services/event-bus-local.ts]:
```suggestion
  private async stageOrEmitEvent(
```

File: packages/core/utils/src/event-bus/index.ts (hunk 1 above)
Comment (carlos-r-l-rodrigues) [reply packages/core/utils/src/event-bus/index.ts]: I believe this should be private, no?

File: packages/core/utils/src/event-bus/index.ts [hunk 4]
```diff
@@ -24,6 +24,15 @@ export abstract class AbstractEventBusModuleService
...
+    eventGroupId: string,
+    eventName: string,
+    data: TData
+  ): Promise<void>
+
+  abstract releaseStagedEvents(eventGroupId: string): Promise<void>
+  abstract clearStagedEvents(eventGroupId: string): Promise<void>
```
Comment (carlos-r-l-rodrigues) [inline packages/core/utils/src/event-bus/index.ts]: Maybe we can find better names here?
I like only `release` and `dismiss`
or `releaseGroupedEvents` and `cancelGroupedEvents`/`dismissGroupedEvents`

**Review by carlos-r-l-rodrigues — APPROVED** [review]

File: packages/core/utils/src/event-bus/index.ts (hunk 1 above)
Comment (riqwan) [reply packages/core/utils/src/event-bus/index.ts]: yeah, lets make this private

Discussed shape on chat. 👍🏻

File: packages/core/utils/src/event-bus/index.ts (hunk 4 above)
Comment (riqwan) [reply packages/core/utils/src/event-bus/index.ts]: can used grouped instead, but wrt to a queue, the action words are release and clear right?

**Review by sradevski — APPROVED** [review]
So far so good! Talking concretely about bullmq, do we have the possibility to group events like this without too much performance overhead?

**Comment by riqwan** [comment]
@sradevski we're pushing it straight to redis, potentially just passing in a luascript instead of performing the actions here directly

File: packages/core/utils/src/event-bus/index.ts (hunk 4 above)
Comment (riqwan) [reply packages/core/utils/src/event-bus/index.ts]: resolved over chat - decision to use group, but the action words the same 💪🏻

**Pull request merged**